File: 2024/6/main.py
import sys

data = open(sys.argv[1]).read().strip()
lines = data.split('\n')

grid = [list(line) for line in lines]

inity, initx = 0, 0
for y, row in enumerate(grid):
    for x, cell in enumerate(row):
        if cell == '^':
            inity, initx = y, x
            break

move = 0

seen = set()

cury, curx = inity, initx
while 0 <= cury < len(grid) and 0 <= curx < len(grid[0]):
    if grid[cury][curx] == '#':
        match move:
            case 0: cury += 1
            case 1: curx -= 1
            case 2: cury -= 1
            case 3: curx += 1
        move = (move + 1) % 4
    seen.add((cury, curx))
    match move:
        case 0: cury -= 1
        case 1: curx += 1
        case 2: cury += 1
        case 3: curx -= 1

ans = len(seen)

ans2 = 0
for blocky, blockx in seen:
    grid[blocky][blockx] = '#'
    cury, curx = inity, initx
    move = 0
    seen = set()
    while 0 <= cury < len(grid) and 0 <= curx < len(grid[0]):
        if grid[cury][curx] == '#':
            match move:
                case 0: cury += 1
                case 1: curx -= 1
                case 2: cury -= 1
                case 3: curx += 1
            move = (move + 1) % 4
        if (cury, curx, move) in seen:
            ans2 += 1
            break
        seen.add((cury, curx, move))
        match move:
            case 0: cury -= 1
            case 1: curx += 1
            case 2: cury += 1
            case 3: curx -= 1
    grid[blocky][blockx] = '.'

print(ans)
print(ans2)

File: 2024/6/main.clj
(require '[clojure.string :as str])

(def data (slurp (first *command-line-args*)))

(def lines (str/split data #"\n"))

(def grid (mapv vec lines))

(def init-coord (some (fn [[y row]]
                        (when-let [x (some (fn [[x cell]]
                                             (when (= \^ cell) x))
                                           (map-indexed  vector row))]
                          [y x]))
                      (map-indexed vector grid)))

(defn turn [cury curx move]
  (case move
    0 [(inc cury) curx (mod (inc move) 4)]
    1 [cury (dec curx) (mod (inc move) 4)]
    2 [(dec cury) curx (mod (inc move) 4)]
    3 [cury (inc curx) (mod (inc move) 4)]))

(defn advance [cury curx move]
  (case move
    0 [(dec cury) curx move]
    1 [cury (inc curx) move]
    2 [(inc cury) curx move]
    3 [cury (dec curx) move]))

(def visited (loop [cury (first init-coord)
                    curx (second init-coord)
                    move 0
                    visited #{}]
               (if (and (>= cury 0)
                        (< cury (count grid))
                        (>= curx 0)
                        (< curx  (count (first grid))))
                 (let [[cury curx move] (if (= (get-in grid [cury curx]) \#)
                                          (turn cury curx move)
                                          [cury curx move])
                       seen' (conj visited [cury curx])
                       [cury' curx' move'] (advance cury curx move)]
                   (recur cury' curx' move' seen'))
                 visited)))

(def ans (count visited))

(def ans2 (->> visited
               (map
                (fn [[blocky blockx]]
                  (loop [ans2 0
                         grid (assoc-in grid [blocky blockx] \#)
                         cury (first init-coord)
                         curx (second init-coord)
                         move 0
                         visited #{}]
                    (if (and (>= cury 0)
                             (< cury (count grid))
                             (>= curx 0)
                             (< curx  (count (first grid)))
                             (= 0 ans2))
                      (let [ans2' (if (contains? visited [cury curx move]) 1 0)
                            [cury curx move] (if (= (get-in grid [cury curx]) \#)
                                               (turn cury curx move)
                                               [cury curx move])
                            seen' (conj visited [cury curx move])
                            [cury' curx' move'] (advance cury curx move)]
                        (recur ans2' grid cury' curx' move' seen'))
                      ans2))))
               (reduce +)))

(println ans)
(println ans2)